osquery/tables/networking/posix/lldp_neighbors.cpp

```diff
@@ -0,0 +1,421 @@
+/*
+ *  Copyright (c) 2014-present, Facebook, Inc.
+ *  All rights reserved.
+ *
+ *  This source code is licensed under the BSD-style license found in the
+ *  LICENSE file in the root directory of this source tree. An additional grant
+ *  of patent rights can be found in the PATENTS file in the same directory.
+ *
+ */
+
+#include <iostream>
+#include <map>
+#include <string>
+#include <vector>
+
+#include <osquery/logger.h>
+#include <osquery/tables.h>
+
+#include <lldp-const.h>
+#include <lldpctl.h>
+
+namespace osquery {
+namespace tables {
+
+/* kNoAtomStrDefault is the default value used when lldp_atom_get_str does not
+ * return a valid string value. */
+const std::string kNoAtomStrDefault = "unknown";
+
+struct ChassisCapability {
+  bool available;
+  bool enabled;
+};
+
+inline std::string getAtomStr(lldpctl_atom_t* atom, lldpctl_key_t key) {
+  if (!atom) {
+    return kNoAtomStrDefault;
+  }
+
+  const char* val = lldpctl_atom_get_str(atom, key);
+  if (val == NULL) {
+    return kNoAtomStrDefault;
+  }
+
+  return val;
+}
+
+inline std::string commaDelimitedStr(lldpctl_atom_t* things,
+                                     lldpctl_key_t key) {
+  if (!things) {
+    return "";
+  }
+
+  std::string result;
+  lldpctl_atom_t* each;
+
+  lldpctl_atom_foreach(things, each) {
+    result = result + getAtomStr(each, key) + ",";
+  }
+  // Remove trailing comma.
+  if (result.size() > 0) {
+    result.pop_back();
+  }
+
+  return result;
+}
+
+class LLDPNeighbor {
+ public:
+  /**
+   * @brief LLDPNeighbor is responsible for retrieving LLDP information on one
+   * neighbor.  On construction, a Row is initialized for member functions to
+   * fill.
+   *
+   * It does not handle the the reference decrementing of the chassis and port
+   * parameter.  That responsibility is on the caller.
+   *
+   * @param port liblldpctl atom type of given port
+   * @param chassis liblldpctl atom type of given chassis
+   *
+   */
+  LLDPNeighbor(lldpctl_atom_t* port, lldpctl_atom_t* chassis)
+      : port_(port), chassis_(chassis) {}
+
+  LLDPNeighbor(const LLDPNeighbor& ln) = delete;
+
+  Row& getNeighbor();
+
+ private:
+  lldpctl_atom_t* port_;
+  lldpctl_atom_t* chassis_;
+  Row row_;
+
+  void getChassis();
+  void getChasisCapability(u_int8_t mask,
+                           std::string availCol,
```

Comment:
Since you're note moving these `std::string`s, pass them by `const std::string& ` to avoid copying the content.